[WIP] l10n_es_aeat_sii: Presentación inmediata de impuesto

acysos opened the pull request:
Hola,

Módulo beta para el SII. En el README están especificadas como utilizarlo y funcionalidades que faltan, tenemos planificado seguir ese orden.
De hecho el primer punto del Roadmap, factura rectificativas ya esta en proceso.

Actualmente conecta perfectamente con hacienda y recupera el mensaje de respuesta de las facturas emitidas y recibidas, tanto de bienes corrientes, servicios e inversión de sujeto pasiva. También comunica correctamente el recargo de equivalencia.

Saludos

acysos committed:
l10n_es_aeat_sii: presentación inmediata de impuesto

jabibi commented on l10n_es_aeat_sii/views/account_invoice_view.xml:
```diff
@@ -0,0 +1,40 @@
+                <notebook position="inside">
+                    <page string="AEAT" groups="account.group_account_manager">
+                        <group>
+	                        <separator string="SII" colspan="4"/>
```
El string podría ir directamente en el group, creo que tiene el mismo efecto, así puedes quitar el nodo separator
Por otro lado el número de columnas por defecto en un group es de 2, por lo que el colspan del separator debería ser 2

(la tabulación no es correcta)

jabibi commented on l10n_es_aeat_sii/views/account_invoice_view.xml:
```diff
@@ -0,0 +1,40 @@
+                <notebook position="inside">
+                    <page string="AEAT" groups="account.group_account_manager">
+                        <group>
+                            <separator string="SII" colspan="4"/>
```
El string podría ir directamente en el group, y se elimina el nodo separator

jabibi commented on l10n_es_aeat_sii/views/account_invoice_view.xml:
```diff
@@ -0,0 +1,40 @@
+        </record>
+
+    </data>
+</openerp>
```
Línea EOF

jabibi reviewed (commented):
He mirado un poco el código sin probarlo, gracias por el trabajo Ignacio.

Los comentarios de código que hago son detalles menores, de poca importancia

oihane commented:
@acysos podéis añadir en este PR en el fichero `requirements.txt`, una línea que ponga `zeep`? De ese modo se tanto el travis como el runbot instalará esa dependencia y se podrá testar.

Commit:
l10n_es_aeat_sii: Carga directa del certificado

AlbertCabedo commented:
únicamente comentar que Hacienda está muy estricta en que coincidan el número de factura, fecha de factura y fecha de contabilización, tal vez para v8 habría que crear el campo fecha de contabilización. En algunas ocasiones he visto que se ha intentado y la mejor opción es que la fecha factura actual sea la fecha de contabilización para que coincida con el periodo y la validación fecha y periodo que se puede activar en los diarios funcione correctamente. La fecha factura si es un campo más informativo para el SII.

acysos committed:
l10n_es_aeat_sii: se añade mapeo de impuestos, traducciones y permisos

acysos committed:
l10n_es_aeat_sii: añadidas facturas de bienes de inversión e intracomunitarias. Se traspasa la configuración del WSDL a parametros

fuentes73 committed:
l10n_es_aeat_sii: Mejoras del certificado + otros

* Ocultar la contraseña del certificado
* Plantilla de cuentas solo obligatorio si "activar SII" activado
* Habilitar un certificado por compañía
* Permisos de acceso
* Tener en cuenta las fechas de los certificados

RamonGuiuGou committed:
l10n_es_aeat_sii. Add connector

fuentes73 committed:
[IMP] l10n_es_aeat_sii: No utilizar los parámetros del sistema y buscar solo certificados activos

acysos committed:
l10n_es_aeat_sii: error en la búsqueda del certificado + correción descripciones en facturas de proveedor

ismaelcj committed:
l10n_es_aeat_sii, refactoring

fuentes73 committed:
l10n_es_aeat_sii: Envío de facturas rectificativas

ismaelcj committed:
l10n_es_aeat_sii, tests

RamonGuiuGou committed:
l10n_es_aeat_sii: Filtros + agrupación:

* Añadir filtros a las facturas para saber si el SII se ha enviado o no
* Corregir la agrupacion de los impuestos si en varias línias de factura hay el mismo impuestos

fuentes73 committed:
l10n_es_aeat_sii: Trabajos readonly, limitar campos, poder re-encolar y retardo configurable

acysos committed:
l10n_es_aeat_sii: copy fix, send button, method send, new filter and translations

Jortolsa-S73 committed:
función para verificar si la factura debe enviarse o no a través del SII, pensada para ser heredada en otros módulos como l10n_es_dua_sii

Jortolsa-S73 committed:
añadidos impuestos de importación al mapeo del SII, cambio de NIF

oihane commented on l10n_es_aeat_sii/tests/test_l10n_es_aeat_sii.py:
```diff
@@ -0,0 +1,193 @@
+        })
+
+    def _open_invoice(self):
+        self.invoice.company_id.sii_enabled = True
```
No habría que añadir algo así
```python
self.invoice.company_id.chart_template_id = self.env.ref(
    'l10n_es.account_chart_template_pymes').id
```

oihane committed:
l10n_es_aeat_sii: update tests

ozono committed:
l10n_es_aeat_sii: función para uso condicional de IDOtro/NIF + inclusión de ImporteTotal

De momento se aplica solo a emitidas

RamonGuiuGou committed:
l10n_es_aeat_sii: Añadir transformación multidivisa en los importes && añadir en la difinicion del producto la configuración de la causa exenta y para poderla enviar a la agencia

oihane committed:
l10n_es_aeat_sii: travis and pylint

pedrorgil committed:
l10n_es_aeat_sii: añadido hook para calcular el precio subtotal por línea de factura. De esta forma permitimos que otros módulos puedan heredar e intervenir en el cálculo del precio subtotal por línea, como por ejemplo para incluir el importe de un tasa, segundos descuentos, etc.

pedrobaeza commented:
Sería mejor no mezcla módulos accesorios en este PR (como el del DUA), ya que este PR es lo suficientemente denso.

Commit:
l10n_es_aeat_sii:

* filtro facturas sii fallidas + clave registro en posición fiscal
* catch job error on invoice

omar7r committed:
[FIX] l10n_es_aeat_sii: En lo posición fiscal ahora se permiten poner dos claves por defecto, una para ventas y otra para compras. También se corrigen los problemas de redondeo si se aplica redondeo global en la contabilización de facturas. [FIX] PEP8. [UPD] Traducciones.

pedrobaeza commented:
He limpiado la rama de commits, agrupando de manera lógica, y he quitado el módulo del DUA de aquí. De esta forma, puedo recuperar el historial de commits para las migraciones a v9 y v10 del módulo, y respetar la atribución de código. 

@Jortolsa, he aislado el módulo del DUA (con los sucesivos commits que teníais incluido el hook y demás) en la rama https://github.com/Tecnativa/l10n-spain/tree/8.0-l10n_es_dua_sii. Puedes coger dicha rama tal cual y ponerla en tu repositorio para hacer el correspondiente PR.

Continúo ahora con los PRs de la 9 y la 10 cogiendo esta base como he dicho.

pedrobaeza commented:
Al final, nadie ha puesto la lista de cosas pendientes. ¿Podemos considerarlo fuera de "Work in progress"?

pedrorgil commented:
Por lo que hemos revisado vemos que queda pendiente:
- El dichoso tema de la fecha de registro, en votación, parece :)
- controlar los impuestos de las facturas emitidas no sujetas. Por ejemplo en las ventas con ISP al 0% no se declara nada y da error

A parte he comprobado que si se envía una factura modificada el SII busca la comunicada anteriormente  por el CIF del emisor + n° factura. Si por ejemplo te equivocas al meter el n° factura de un proveedor y la comunicas, como vuelvas a borrador, corrijas el n° de factura y vuelvas a enviar (como modificada A1) da error porque dice que el registro no se encuentra. Si vuelves a enviar de nuevo ahora ya la envía como A0 (enviada SII ahora es falso) y crea una nueva, con lo que tienes una factura contabilizada y dos comunicadas (con dos n° de factura distintos). Esto puede ser algo "normal" que pase, en esos casos ¿que hacer? Porque la que se puede liar es gorda y hacer un abono no es viable en este caso. ¿Alguien ha lidiado con esto? Lo que se me ocurre es que haya que cancelar la factura con el número mal, comunicar la baja y dejarla cancelada y crear una nueva con el número de factura correcto, por lo que no se debería ni siquiera dejar modificar el número de factura de una factura de proveedor ya comunicada.

pedrobaeza commented:
Gracias por el resumen. Supongo que el know issues del README también aplica como pendiente, aunque algunas cosas sean para módulos satélite. Por el momento, con el tema conector y un retardo establecido de algunas horas, esto no debería ser especial problema para la mayoría de casos, ya que se cancelaría el trabajo en lugar de enviar otra. Para el resto de casos, es probar si la cancelación de la original funciona.

pedrorgil commented:
Voy a hacer alguna prueba del servicio de cancelación. Creo que sería conveniente habilitar una opción para las facturas canceladas que permita comunicar la baja de forma manual, como botón "Baja SII" que aparezca solo en estado cancelada. Si lo veis conveniente y funciona haré un PR.

pedrobaeza commented:
Voy a pasar a hacer una refactorización del módulo, que incluye:

* Solucionar inconsistencias de estilo
* Reducir la complejidad ciclomática (https://es.wikipedia.org/wiki/Complejidad_ciclom%C3%A1tica)
* Añadir salvaguardas en el acceso a ciertos datos
* Añadir otras optimizaciones

Por lo que si tenéis algo pendiente que proponer, por favor esperaos a que haya hecho esos cambios para basaros en la nueva versión (las haré a lo largo del fin de semana + lunes).

pedrobaeza committed:
[RFR] l10n_es_aeat_sii: General refactoring:

* Onchanges renamed for avoiding conflicts with other modules
* Methods renamed for avoiding conflicts with other modules
* Some method docstrings.
* Use recordset instead of lists.
* Remove incorrect mix of @api.multi and @api.model decorators
* Remove method arguments that are passed when there are other data that are taken from self.
* Place in the proper model some of the methods for using self reference
* Variables renamed according their use.
* POT file removed.
* Reduce cyclomatic complexity of some methods.
* Code style more homogeneous.
* Uniformize file headers.
* Full Spanish translation.
* Add icon.

pedrobaeza commented:
Bueno, ya he realizado la refactorización general con los siguientes cambios:

- Renombrados onchange para evitar choques con otros módulos
- Renombrados métodos para evitar choques con otros módulos
- Docstring de métodos
- Utilizar recordset en lugar de listas
- Quitar mezcla incorrecta de decoradores @api.multi, @api.model
- Parámetros pasados por argumento incoherentes con luego el resto de código que toman datos de self.
- Colocar los métodos en su sitio adecuado para aprovechar las referencias self.
- Nombres de variables acorde a su contenido.
- Archivo POT no necesario.
- Eliminada complejidad ciclomática de algunos métodos.
- Estilo de código más homogéneo.
- Cabeceras de archivo uniformizadas.
- Traducción al español completa.
- Añadido icono.

Hay un cambio que me tenéis que confirmar si funciona, ya que todavía no tengo acceso a un sistema de prueba, que es el que en lugar de ir preguntando cada vez si está ya presente la clave X del diccionario, lo que se hace es construir un diccionario base con todas esas claves. Es mucho menos código, pero la AEAT debe admitirlo.

Hay varias cosas que no tengo claras de dejar así, aunque no quería tampoco iniciar una revolución metiendo los cambios:

* El que el conector sea opcional. Ahora mismo hay código y vistas incoherente si se usa el conector y viceversa. Puesto que el conector es una dependencia en el manifiesto, y se puede programar cuándo realizar el envío (o incluso se puede desactivar la ejecución de trabajos y que se haga manual vía la tarea planificada), debe quitarse la opción manual.
* @RamonGuiuGou, añadiste una opción para escoger el tipo de exención en productos, pero no hay más código que utilice ese mapeado. ¿Vas a añadir ese código? Si no, se debería quitar ese mapeado.
* Sobre el gestor de certificados, no termina de gustarme que se almacene en esa ubicación arbitraria, que además está fuera de cualquier sistema de backup actual, y además puede estar restringida en entornos de despliegue automáticos. ¿Cuál es la razón para haberlo hecho así y no guardar en BD (tal vez encriptado)?
* Esto es más una nota de aviso para tod@s con las traducciones: siempre se utiliza trato de usted y además no podéis poner un texto directamente en español y no poner la cadena como unicode (véase "Compañía" en el gestor de certificados). Para más referencia, ver https://odoospain.odoo.com/blog/noticias-odoospain-1/post/traduccion-de-odoo-al-espanol-de-espana-3

pedrorgil commented:
@pedrobaeza gracias por el trabajo!! Mañana probaremos lo del diccionario de claves para co firmar que funcione.

En cuanto a que el conector no sea opcional no me parece mal, pero no sé si te he entendido bien en cuanto a la opción "manual". Yo sí mantendría la opción de presentación manual (que haya que darle al botón) aunque se haga a través del conector.

pedrobaeza commented:
Sí, también digo de eliminar la opción manual, ya que si lo quieres manual, siempre puedes utilizar las opciones del conector para gestionar qué enviar.

pedrorgil commented:
Pero eso obliga a que el contable sea consciente de lo que es el conector y cómo utilizarlo, y como de todo hay en la viña del señor seguro que habrá muchos que no querrán ni saber qué es, "tú déjame que yo presente lo que quiera cuando yo quiera, no cuando le de al validar" dirán, como si los viera vamos... por eso creo que la opción manual no molesta, puede respetar el uso del conector pero da la posibilidad de que cada contable presente lo que quiera cuando quiera, ya es responsabilidad de cada uno que no se le quede ninguna factura sin presentar en plazo si deciden utilizar la opción manual, que muchos estoy seguro que al menos al principio la querrán.

pedrobaeza commented:
Es que el contable no debe decidir eso. Lo debe decidir el responsable financiero. Yo no voy a estar pendiente de que un becari@ que mete facturas le dé a ese botón. Siempre tienes el control desde la pestaña SII, donde ves los trabajos, por lo que no tienen que ser tampoco conscientes del conector. Es más, no sé ni cómo plantéais con la responsabilidad que implica por vuestra parte el dejar a alguien que pulse cuando le dé la gana...

pedrorgil commented:
Todo eso depende de si hay un becario, un responsable financiero o solo un contable-responsable financiero y becario :). Si el que manda te dice que el quiere darle al botón, a ver quién es el guapo que le quita la idea. Y ya estamos con el módulo personalizado que lo habilita... solo digo que como o molestarte si decisión de cada uno, por qué quitarlo

pedrobaeza commented:
Ya lo he dicho, porque es más código a mantener (y quitar errores) y hay ahora mismo incoherencias, como la posibilidad de darle a enviar varias veces. Además, he visto que hay un control para mientras se está enviando, pero no para cuando ya se ha enviado el trabajo.

pedrorgil commented:
Más código a mantener, es solo comprobar la configuración al validar y si esta manual no hacer nada. El botón de enviar llama al mismo método que al validar, con lo que es el mismo código. Lo de enviar varias veces lo evitamos con poner invisible el botón de enviar si ya está enviada la factura, se le añade en attrs a la vista y listo. 

No quiero polemizar más, entiendo que a ti te pueda sobrar, pero se que es una funcionalidad que van a pedir porque al menos a nosotros ya nos lo han pedido muchos los clientes, porque al principio van a estar "cagados" y no quieren delegar en que se haga automático con o sin retardo. Pero lo dicho, sin entrar en más polémica, lo que se decida bien está y si es necesario ya haremos nosotros una adaptación.

pedrobaeza committed:
[FIX] l10n_es_aeat_sii:

* Falla la variable TipoImpositivo de la línea 281, incorrecta se manda como key no como valor.
* Falla el recargo de equivalencia en las facturas enviadas
* Añade el recargo de equivalencia, cuando hacienda solo lo pide si es necesario, este error estaba corregido y ha vuelto. Si no lo tiene recargo no hay que ponerlo. Error en línea 730

pedrobaeza commented:
Ya están corregidos los problemas mencionados por @acysos:

* Falla la variable TipoImpositivo de la línea 281, incorrecta se manda como key no como valor.

Este problema viene originalmente del último commit de @omar7r (https://github.com/acysos/l10n-spain/pull/31/files#diff-1e50d7d6033ffb0d15c5823d421fc2c2R381), que yo me he fiado de que era una clave existente, pero en realidad no está en ningún sitio

* Falla la función, lineas 279 y 284, no se contrala si existen lineas de impuestos.

Lo bueno del método seguido es que no hay que comprobar si hay líneas de impuestos. Si no las hay, establece una lista vacía. Sí que está por probar lo que decía de si Hacienda permite esa lista vacía. Si no, se soluciona rápidamente.

* Falla el recargo de equivalencia en las facturas enviadas

Ya está reparado el código, que además en la versión de Ignacio tiene un error, ya que sumaba el tipo impositivo (0,5%, 1,4%, etc).

* Añade el recargo de equivalencia, cuando hacienda solo lo pide si es necesario, este error estaba corregido y ha vuelto. Si no lo tiene recargo no hay que ponerlo. Error en línea 730

Ya he puesto para que no lo rellene por defecto. Sólo lo pone cuando se detecta recargo de equivalencia.

pedrobaeza commented:
@pedrorgil lo que me refiero de más código a mantener es más condiciones de UI a controlar. Fíjate lo que ocurre ahora mismo: tienes un botón para enviar al SII, junto con una acción de servidor para hacerlo masivamente, pero en cambio ese botón no se esconde en caso de que se haga automáticamente (ni aunque esté enviado), lo que hace que puedas volver a enviarlo las veces que quieras, encolando trabajos para lo mismo con errores que serán más difíciles de explicar, porque de hecho machacará la respuesta correcta que había anteriormente en los campos `sii_csv`, `sii_return`, etc.

Por eso digo que veo más fácil explicar el manejo de los trabajos del conector y realizar adaptaciones en torno a ello que tener la doble opción. No digo de decirles a los clientes que no envíen ellos cuando quieran (que sigo pensando que es peligroso, aunque no es mi lucha), pero que se puede hacer tan fácil como no iniciar el jobrunner del conector (no poniendo lo de `--load=web,web_kanban,connector`), y poniendo un botón de enviar en la pestaña SII que lo que haga sea procesar el trabajo vía el conector "forzándolo" en ese momento.

¿Me entiendes por dónde voy?

pedrobaeza committed:
[FIX] l10n_es_aeat_sii: Greenify tests + warning

oihane commented on l10n_es_aeat_sii/tests/test_l10n_es_aeat_sii.py:
```diff
@@ -138,16 +141,20 @@ def _get_invoices_test(self, invoice_type, special_regime):
+                    self.env['account.invoice']._get_in_taxes_basic_dict()
+                ),
                 "FechaRegContable":
                     datetime.strptime(
```
Aquí creo que también puedes hacer el cambio de usar el field.Date.from_string para hacer la conversión.

pedrobaeza replied:
Efectivamente, se podría cambiar a `fields.Date.from_string`. Reviso más posibles ocurrencias de este tipo y las cambio.

pedrobaeza committed:
[FIX] l10n_es_aeat_sii: Don't create basic structure

AEAT doesn't allow to send empty structures, so it's built on demand.

oihane commented on l10n_es_aeat_sii/models/account_invoice.py:
```diff
@@ -0,0 +1,674 @@
+                #     'account.fp_intra').id:
+                #     res = serv.SuministroLRDetOperacionIntracomunitaria(
+                #         header, invoices)
+                if res['EstadoEnvio'] == 'Correcto':
```
que pasa con el estado del envío `'AceptadoConErrores'`?

pedrobaeza replied:
Ahora mismo en ese caso no se da por enviado, pero se registra la respuesta. ¿Te has encontrado ese caso? ¿Permite luego volver a enviar la factura tal cual?

acysos commented:
En la última versión está corregido

pedrobaeza commented:
@acysos a qué te refieres que está corregido y cuál última versión?

amcor commented:
Buenas @pedrobaeza, no se si lo has visto ya, pero el principal problema son los diccionarios de keys, por ejemplo, _get_in_taxes_basic_dict https://github.com/acysos/l10n-spain/blob/8.0-l10n_es_aeat_sii/l10n_es_aeat_sii/models/account_invoice.py#L291 incluye la key de InversionSujetoPasivo, cuando no se encuentra ningún impuesto manda el InversionSujetoPasivo\DetalleIVA pero sin ninguna ocurrencia, y ya el ws de hacienda casca.

 _Expected at least 1 items (minOccurs check) 
(SuministroLRFacturasRecibidas.RegistroLRFacturasRecibidas.FacturaRecibida.DesgloseFactura.InversionSujetoPasivo.DetalleIVA)_

Más arriba creo que preguntabas si esos diccionarios funcionaban, pues parece ser que no :(

Un saludo.

pedrobaeza commented:
OK, es la confirmación que necesitaba. Ya tenía montado casi mi sistema de pruebas, pero con eso dicho, rápidamente lo cambio y aunque el código no quede tan limpio, sí que reduzco de todas formas un poco la complejidad ciclomática con el método `setdefault`. Ahora os aviso en cuanto lo tenga.

pedrobaeza commented:
Pues nada, ya está rehecho no precreando el diccionario y creando las claves bajo demanda. Probad por favor a ver si encontráis cualquier otro problema.

amcor commented on l10n_es_aeat_sii/models/account_invoice.py:
```diff
@@ -191,126 +191,95 @@ def _get_sii_header(self, tipo_comunicacion):
                 if tax_line in (taxes_sfess + taxes_sfesse):
+                    type_breakdown = taxes_dict.setdefault(
+                        'DesgloseTipoOperacion', {
+                            'PrestacionServicios', {'Sujeta': {}},
```
Cuidado con el diccionario, se ha quedado una "," en lugar de los ":".
Corregido eso, facturas normales ya están cargando ;)

pedrobaeza replied:
Cierto! Gracias por pillarlo.

pedrobaeza replied:
Cambiado

RamonGuiuGou commented:
Nos hemos bajado el último commit (el 5837f51), y estamos haciendo pruebas... a nosotros aún nos da problemas al enviar una factura de cliente hicienda devuelve un error "El campo TipoImpositivo debe estar informado siempre que la ClaveRegimenEspecialOTrascendencia no tenga un valor 03, 05, 16 o alguna ClaveRegimenEspecialOTrascendenciaAdicional este informada".... , hemos subido un PR el 33 con la intención de corregir esto y otras cosas. 
El commit 5837f51 és el último? O hay otra rama más actualizada?

pedrobaeza commented:
Ése es el último commit. Reviso entonces vuestro PR.

RamonGuiuGou committed:
[FIX] l10n_es_aeat_sii: Corrección de pequeños problemas

* TipoImpositivo
* Default del refund_type
* Clave de registro facturas proveedor

pedrobaeza commented:
Añadido nuevo commit con los cambios de Ramón para esos casos comentados.

pedrobaeza committed:
[FIX] l10n_es_aeat_sii: Protect correctly OpenSSL

pedrobaeza committed:
[FIX] l10n_es_aeat_sii: Correct Python dependency name

amcor commented:
He continuado realizando pruebas, las facturas ISP no se mandan bien, no acaba de construirse la estructura necesaria.
Msg 'Las operaciones podr\xe1n tener parte sujeta y parte no sujeta. Por tanto, puede aparecer solo un bloque o ambos, pero al menos debe aparecer uno (Sujeta y/o No sujeta)'

Creo que en PR de @ozono se incluye ya corregido, por lo menos, lo referente a esta parte.

Un saludo

pedrobaeza commented:
OK, ahora voy a pasar a integrar ese PR y os aviso.

pedrobaeza committed:
[FIX] l10n_es_aeat_ssi: Translations

RamonGuiuGou commented:
He hecho un PR el 34 añadiendo el tema de las causas exentas configurado en los productos y la corrección de las facturas de proveedor intracomunitarias que daba un error. Eso lo he transformador de un PR de un compañero a la nueva versión. Faltaría provarlo un poco más por si me he dejado algo.
Me voy a poner con las facturas intracomunitarias de cliente que daba algún pequeño problema.

ozono committed:
[FIX+IMP] l10n_es_aeat_sii: Facturas exentas + desglose emitidas + ImporteTotal + otros

* Ajuste DesgloseFactura/DesgloseTipoOperacion en emitidas

  Evita/adapta que se incluyan DesgloseFactura/DesgloseTipoOperacion (excluyentes).
  Si es obligatorio usa DesgloseTipoOperacion para ciertas operaciones.
  Se hace al final de _get_sii_out_taxes para tocar lo menos posible lo que ya hay.

* Mapeado S_IVA0_E en SFESBE
* Se añade a recibidas clave ImporteTotal.

pedrobaeza commented:
Ya están mergeados todos los cambios de operativa en la nueva versión. Por favor probad que funciona y si dais vuestro visto bueno, deberíamos mergear el PR para contar con una primera versión para ya realizar el resto de PRs directamente a OCA/l10n-spain y contar con el apoyo de Travis y Runbot.

RamonGuiuGou committed:
[FIX]l10n_es_aeat_sii. IR a buscar el TipoImpositivo a los impuestos
hijos de los intracomunitarios. Para no presentar el 100% del padre.

RamonGuiuGou commented:
He subido un par de commits más, para intentar solventar algunos pequeños detalles en las facturas intracomunitarias.

pedrobaeza commented:
Dónde, Ramón? Si es en el PR abierto, te rogaría que lo limpies, ya que aparecen 137 commits!

RamonGuiuGou commented:
@pedrobaeza Si me sale todo el historico de commits en el PR... y no ser por que. Como se pude limpiar con un rebase? Hay algún manual Github OCA?

oihane commented:
@RamonGuiuGou https://www.devroom.io/2011/07/05/git-squash-your-latests-commits-into-one/ puede que te sirva.

pedrobaeza commented:
No, más bien sería un manual GIT genérico. Prueba a hacer un `git rebase origin/8.0`, donde origin es el remote de OCA/l10n-spain. Si se generan conflictos, lo mejor será que empieces de esta rama del PR, y hagas cherry-pick de los commits que quieras.

RamonGuiuGou commented:
Gracias a los dos! Ahora nos hemos bajado este PR y hemos solucionado la limpieza de la rama pasando con cherry-pick. Pero, nos encontramos que no tenemos permisos para subir a este PR con mi usuario.
Como lo podemos solucionar?

pedrobaeza commented:
Lo que puedes hacer es subirlo a la rama original del PR con `git push -f` para sobreescribir y ya lo cojo de ahí, o bien me indicas qué rama es y hago igualmente yo el cherry-pick a esta rama.

ozono commented:
Un comentario a tener en cuenta: ojo con el signo de las rectificativas.  Hay que comunicarlas en negativo cuando en Odoo aparecen en positivo.

RamonGuiuGou commented:
@pedrobaeza con el git push -f en la rama original nos da el error igual de permisos. Los commits son 92bb556 y 17add50 estan en el repositorio minorisa/l10n-spain rama 8.0-sii, cuando puedas si los puedes coger corrigen pequeños detalles.

@ozono consultando el FAK de http://www.agenciatributaria.es/static_files/AEAT/Contenidos_Comunes/La_Agencia_Tributaria/Modelos_y_formularios/Suministro_inmediato_informacion/FicherosSuministros/V_05/FAQs_v05.pdf respuesta 2.8 por facturas de sustitución hay dos opciones, la que comentas con importes negativos dónde se deberían mandar dos facturas y la opción 1 dónde es el método que utilizamos y se tiene que mandar en positivo. En mi opinión, si no lo hemos entendido mal cumplimos la normativa, a nosotros nos falta terminarlo de probar para que no se nos escape algún detalle.

ozono commented:
@RamonGuiuGou: pero también hay el caso de las facturas rectificativas por diferencias, en las que no hay otra opción que mandar el importe en negativo. Ver respuesta 2.9 y ejemplos en documentación técnica.

ozono commented:
@pedrobaeza: tengo un par de ajustes para el modo en que se están registrando las facturas después de tu refactorización. Por ejemplo, para intracomunitarias emidas se está mandando el bloque 'Exenta' al mismo tiempo que el 'NoExenta' Cual es el procedimiento recomendable ahora para proponer los cambios?

pedrobaeza commented:
@ozono pues puedes subir ese cambio a tu rama y mencionar aquí el commit (enlace) y así lo integro. Por eso quiero mergear lo antes posible cuando lo tengamos lo suficientemente estable para que ya a partir de ahí se hagan PRs directos a l10n-spain.

ozono committed:
[FIX] l10n_es_aeat_sii: NoExenta solo cuando corresponda y condiciones

ozono commented:
@pedrobaeza El commit con algunos ajustes para las emitidas es https://github.com/ozono/l10n-spain/commit/5fd19c27181fb7fb5fb638cd8780addf46d76ea8

pedrorgil commented:
@ozo esos son los cambios que también habíamos hecho nosotros, más o menos igual. Reviso vuestro commit y si no veo nada distinto al nuestro esperamos entonces a que hagáis PR vosotros.

Con las prisas estamos empezando a pisarnos, deberíamos organizarnos y al menos poner por aquí lo que cada uno esté haciendo para no duplicar trabajo

pedrobaeza committed:
[IMP] l10n_es_aeat_sii: Several improvements:

- Cambio del campo "Carpeta" al asistente de obtención de claves.
- Incluir el tipo de rectificación en el asistente de factura rectificativa.
- Renombrados dos campos que faltaban sin prefijo `sii_` para evitar colisiones con otros módulos.
- Solventadas algunas inconsistencias del UI de la factura.
- Campo readonly para related para evitar efectos laterales
- PEP8

JuanjoA committed:
[IMP] l10n_es_aeat_sii: Segregación de lógica en _get_sii_invoice_dict

pedrobaeza committed:
[FIX+IMP] l10n_es_aeat_sii: Better tax dict treatment + PEPP8

pedrobaeza commented:
Ya está integrado todos los commits/PRs pendientes + otra vuelta que le he dado a las siguientes cosas:

- Cambio del campo "Carpeta" al asistente de obtención de claves.
- Incluir el tipo de rectificación en el asistente de factura rectificativa.
- Renombrados dos campos que faltaban sin prefijo `sii_` para evitar colisiones con otros módulos.
- Solventadas algunas inconsistencias del UI de la factura.
- Campo readonly para related para evitar efectos laterales.

Jortolsa-S73 commented:
@pedrobaeza por nuestra parte hemos hecho pruebas en varios clientes y consideramos que está bastante maduro para incluirlo en el repositorio.

pedrorgil committed:
[FIX] Cambiado uso directo del campo price_subtotal de invoice.line por el método "_get_sii_line_price_subtotal" definido para el cálculo de este dato.

ozono committed:
[FIX] l10n_es_aeat_sii: float_round para redondear resultados

- Se usa float-round en el redondeo
- También se redondean los valores en taxes_to
- Se incluye el redondeo de la BaseImponible
- Se aplica también a recibidas

pedrorgil committed:
[IMP] l10n_es_aeat_sii: Anulación de facturas + estados + otros:

* Servicio de anulación de facturas presentadas al SII
* Nueva forma de control de estado de presentación SII en base a estados.
* Control de visualización de botones.
* Hook para cálculo de fecha de registro contable de facturas recibidas.

pedrorgil commented:
He realizado un PR incluyendo el servicio de anulación de facturas en el SII https://github.com/acysos/l10n-spain/pull/38

Por favor revisadlo para que podamos seguir avanzando.

RamonGuiuGou commented:
@pedrobaeza hemos estado haciendo pruebas de facturas de proveedor y hemos encontrado un pequeño bug si pones más de un impuesto en la factura. Tal como esta ahora hicenda solo coge el primer impuesto.
Para solucionarlo se ha de modificar esto:
        if taxes_f:
            taxes_dict.setdefault(
                'DesgloseIVA', {'DetalleIVA': [taxes_f.values()]},
            )  
por 
            taxes_dict.setdefault(
                'DesgloseIVA', {'DetalleIVA': taxes_f.values()},
            )
función _get_sii_in_taxes. En formato lista no le gusta...
Si lo quieres modificar tu mismo, o sino te hacemos el commit, pero como es un pequeño detalle y el commit lo tienes que coger de nuestra rama posiblemente si lo cambías iras más rápido.

JuanjoA committed:
[IMP] l10n_es_aeat_sii: Añadido check para activar o no SII según posición fiscal

pedrobaeza committed:
[FIX] l10n_es_aeat_sii: Don't populate list on list on 'DetalleIva'

pedrobaeza commented:
Ya están todos los PRs (excepto el de anulación SII, que he hecho comentarios) + el cambio que mencionas, Ramón, mergeados.

pedrobaeza commented:
Por cierto, en los PRs que hagáis a la rama de acysos/l10n-spain, por favor mencionadme directamente para que me entere.

omar7r commented:
@pedrobaeza Un nuevo PR https://github.com/acysos/l10n-spain/pull/41 que añade algunas mejoras y correcciones

Por otra parte, también he añadido un punto en el "Known issues" porque en el FAQ del SII indican que los importes tienen que ir siempre en €, sería necesaria una refactorización general para incluir esto.

> 2.36. ¿Puede registrarse la Base Imponible, Cuota del impuesto e Importe total de la
factura en divisas?
No. Debe consignarse su contravalor en euros, aplicando el tipo de cambio vendedor
fijado por el Banco de España vigente a la fecha de devengo de la operación (art. 79
once Ley 37/1992). 

Y veo que se están utilizando los campos amount_total y amount_tax que aunque los pasemos a los que añade el módulo "account_invoice_currency" se estarían dando importes con IRPF incluído o con impuestos no sujetos en casos especiales, por ejemplo IGIC, si la empresa española tiene almacenes en canarias por ejemplo, creo que estos campos deberían obtenerse en una función en la que se pudieran gestionar casuísticas o de una suma general de los desgloses que se presenten.
No se que opináis.

cincmi commented:
cuando se hará el merge a la localización española? para tener ya una versión estable...yo tenia un PR sobre acysos https://github.com/acysos/l10n-spain/pull/30 que no sé si alguien se lo ha mirado...

pedrobaeza commented:
Estoy revisando ahora mismo lo último que puso @pedrorgil el fin de semana y @omar7r y hoy mergeo sin falta. @cincmi tu PR luego lo propones directamente a OCA/l10n-spain una vez hagamos este merge.

cincmi commented:
@pedrobaeza ok

pedrorgil committed:
[FIX] Corregido metodo _get_sii_line_price_subtotal que debía devolver el subtotal (multiplicando por la cantidad) ya que se utiliza para la suma de bases imponibles en algunos sitios, y se ha añadido un nuevo método para cálculo de precio unitario efectivo de las líneas que se utiliza en la llamada al método _compute_all de account.tax.

oihane commented on l10n_es_aeat_sii/data/aeat_sii_mapping_registration_keys_data.xml:
```diff
@@ -0,0 +1,146 @@
+            <field name="code">15</field>
+            <field name="name">Factura con IVA pendiente de devengo - operaciones de tracto sucesivo</field>
+            <field name="type">sale</field>
+        </record>
```
Faltaría la clave para las facturas del primer semestre
```xml
            <field name="code">16</field>
            <field name="name">Primer semestre 2017</field>
            <field name="type">sale</field>
```

oihane commented on l10n_es_aeat_sii/data/aeat_sii_mapping_registration_keys_data.xml:
```diff
@@ -0,0 +1,146 @@
+            <field name="code">13</field>
+            <field name="name">Factura correspondiente a una importación (informada sin asociar a un DUA)</field>
+            <field name="type">purchase</field>
+        </record>
```
Faltaría la clave para las facturas del primer semestre
```xml
            <field name="code">14</field>
            <field name="name">Primer semestre 2017</field>
            <field name="type">purchase</field>
```

pedrorgil commented:
He creado un PR que controla el signo de las facturas rectificativas por diferencias, como bien apuntaba @ozono en este comentario anterior https://github.com/OCA/l10n-spain/pull/464#issuecomment-310027994

igallart commented:
Hola, he tenido un problema con una factura y los decimales. Resulta que una factura con dos líneas, donde el precio unidad es 93.8250 y la otra 900, el campo inv_dict['FacturaExpedida']['TipoDesglose']['DesgloseTipoOperacion']['Entrega']['Sujeta']['Exenta']['BaseImponible'] es 993.825, por lo que hacienda me devuelve tipo incorrecto de baseImponible. Si cambio el valor a 993.83 si que funciona.

¿Es un error de programación o es que tenemos algo mal?
Gracias.

igallart commented:
Me contesto a mí mismo. Acabo de actualizarme el código y se me ha funcionado la misma factura. Gracias.

pedrobaeza commented:
Por lo que sea los últimos commits no están entrando en este PR, así que he tenido que subir la rama a mi organización, y hacer el PR #530.

Cierro entonces éste.

pedrobaeza commented:
Bueno, tras hacer las últimas comprobaciones, y aunque siempre puede añadirse más y más, procedo a mergear el otro PR con los últimos cambios para ya contar con una versión lo suficientemente estable como para tenerla de base. Posibles fallos o mejoras ahora ya se podrán hacer directamente como PR a esta base en OCA/l10n-spain.

Gracias a tod@s por la colaboración en el desarrollo y testeo del sistema.
